File: SelectionCode.m
clear all
close all

fmat = xlsread('GP2017a','Point Allocations'); %should be Nxk matrix where N is number of students and k is num of projects
B = xlsread('GP2017a','Project Maximums'); %should be kx1

k = size(B,1); %The number of projects
N = size(fmat,1);
if size(fmat,2) ~= k
    disp('Matrices are not entered correctly')
end

f = -reshape(fmat',N*k,1);
Beq = ones(N,1);
Aeq1 = [ones(1,k) zeros(1,N*k)];
Aeq2 = repmat(Aeq1,1,N);
Aeq3 = Aeq2(1:N*N*k); 
Aeq = reshape(Aeq3,N*k,N)';

A0 = [1 zeros(1,k-1)];
A1 = [repmat(A0,1,N) 0];
A2 = repmat(A1,1,k);
A3 = A2(1:N*k*k);
A = reshape(A3,N*k,k)';

numVars = N*k;
intcon = 1:numVars;
lb = zeros(numVars,1);
ub = ones(numVars,1);

[X Fval Exit] = intlinprog(f,intcon,A,B,Aeq,Beq,lb,ub);
Solution = reshape(X,k,N)'
-Fval
Exit

xlswrite('GP2017a.xlsx',Solution,'Solution','D3')

File: SelectionCode.R
### Replicating the costellization function in R

## Read the data

library(readxl)
library(tidyverse)
library(lpSolve)

file_path <- "GP2017jc.xlsx"

points <-  read_excel(file_path, sheet = "Point Allocations", col_names = T)

fmat <- dplyr::select(points, -c(1,2)) %>% 
  as.matrix()

B <-  read_excel(path = file_path, sheet = "Project Maximums", col_names = T) %>%
  select(Points) %>% 
  as.matrix()

k <- dim(B)[1] # Number of projects
N <-  dim(fmat)[1]
if (dim(fmat)[2] != k){
  print('Matrices are not entered correctly')
}

f <- t(fmat) %>% 
  as.data.frame() %>% 
  gather(Project, Points) %>% 
  select(Points)

A0 <- diag(1, k)
A <- A0

for (i in 1:(N-1)){
  A <- cbind(A, A0)
}

Aeq <- matrix(0, N, N*k)

for (i in 1:N) {
  Aeq[i, (i*k-k+1):(i*k)] = 1
}

Beq <- matrix(1, N, 1)

const.mat <- rbind(A, Aeq)
const.dir <-  c(rep("<=", k), rep("==", N))
const.rhs <- rbind(B, Beq)

solution <- lp(direction = "max", objective.in = f, const.mat = const.mat, const.dir = const.dir, const.rhs = const.rhs, all.bin = T)

sol_out <- cbind(points$`Last Name`, points$`First Name`, matrix(solution$solution, N, k, byrow = T))
colnames(sol_out) <- colnames(points)

write.csv(x = sol_out, file = "Solutions.csv", row.names = F)


